components/client/src/test/unit/java/com/hotels/styx/client/TransportTest.java

```diff
@@ -67,111 +70,150 @@ public void setUp() {
+        subscriber.unsubscribe();
 
-        verify(pool).borrowConnection();
-        verify(connection).write(any(HttpRequest.class));
-        verify(pool).closeConnection(any(Connection.class));
+        response2.consume();
+
+        verify(pool).returnConnection(any(Connection.class));
     }
 
     @Test
-    public void releasesConnectionWhenRequestFailsAfterHeaders() {
-        ConnectionPool pool = mockPool(mockConnection(responseProvider));
+    public void releasesIfRequestIsCancelledBeforeHeaders() {
+        Connection connection = mockConnection(PublishSubject.create());
+        ConnectionPool pool = mockPool(connection);
 
         transport.send(request, Optional.of(pool), APP_ID)
                 .response()
-                .subscribe(new TestSubscriber<>());
-
-        responseProvider.onNext(response);
+                .subscribe(subscriber);
 
-        verify(pool, never()).returnConnection(any(Connection.class));
-        verify(pool, never()).closeConnection(any(Connection.class));
+        assertTrue(subscriber.getOnNextEvents().isEmpty());
+        assertTrue(subscriber.getOnCompletedEvents().isEmpty());
+        assertTrue(subscriber.getOnErrorEvents().isEmpty());
 
-        responseProvider.onError(new RuntimeException("Connection failed - after the headers"));
+        subscriber.unsubscribe();
 
         verify(pool).closeConnection(any(Connection.class));
     }
 
     @Test
-    public void releasesIfRequestIsUnsubscribedBeforeHeaders() {
+    public void ignoresRequestCancellationAfterHeaders() {
+        // Request observable unsubscribe/cancel has to be ignored after "onNext" event.
+        // This is because Reactor Mono will automatically cancel after an event has
+        // been published.
+
+        TestPublisher<Buffer> testPublisher = TestPublisher.create();
         Connection connection = mockConnection(responseProvider);
         ConnectionPool pool = mockPool(connection);
 
-        HttpTransaction transaction = transport.send(request, Optional.of(pool), APP_ID);
+        transport.send(request, Optional.of(pool), APP_ID)
+                .response()
+                .subscribe(subscriber);
 
-        transaction.response().subscribe(subscriber);
+        responseProvider.onNext(HttpResponse.response(OK).body(new ByteStream(testPublisher)).build());
+
+        assertEquals(subscriber.getOnNextEvents().size(), 1);
+        assertTrue(subscriber.getOnCompletedEvents().isEmpty());
+        assertTrue(subscriber.getOnErrorEvents().isEmpty());
 
         subscriber.unsubscribe();
-        verify(pool).closeConnection(any(Connection.class));
+
+        verify(pool, never()).closeConnection(any(Connection.class));
+        verify(pool, never()).returnConnection(any(Connection.class));
     }
 
     @Test
-    public void releasesIfRequestIsUnsubscribedAfterHeaders() {
+    public void returnsConnectionBackToPool_delayedResponseError() {
```

Comment:
Another test that does not follow naming conventions.

Reply:
We should discuss and clarify our naming conventions. Could you clarify what aspect doesn't conform to naming conventions? 

In the meanwhile I will rename this test to: `returnsConnectionBackToPoolDueToCancelledHeaders`.